### DISCUSSION TITLE
Need help to compile on msys2

### DISCUSSION THREAD

**Discussion by Kosette** [Rafostar/clapper#450] [General]
Can anyone provide some advices on compiling libclapper and libclapper-gtk on msys2? 

I'm trying to compile a program dependent on clapper-rs to Windows, but it prompted missing clapper and clapper-gtk during compilation. However, I did not find the relevant dependency packages under mingw64, so I am considering compiling them manually.

But I've run into an issue. 
```
D:\a\clapper\clapper\builddir/../src/lib/gst/plugin/handlers/gl/gstclapperglcontexthandler.c:457:(.text+0xb8c): undefined reference to `gst_gtk_invoke_on_main' 
```

I installed the dependencies related to gstreamer in the mingw64 environment, installed meson and ninja, but I am very unfamiliar with them, so I don't know whether the error message is missing the dependencies or what I did wrong.

**Comment by Rafostar** [comment]
> missing clapper and clapper-gtk during compilation

These are recently introduced libraries that come from building Clapper (can be built without the app if needed too). But these things are SO new, that I haven't tried building on Windows yet. Supporting Windows was never a requirement for a release of Clapper 0.6.0 version. But if there are users/apps/need for them, I will see what can be done.

There is already an open feature request for adding windows support (https://github.com/Rafostar/clapper/issues/103) and at that time I already succeeded in getting video to work on Windows (a.k.a the hardest part), so what is left should be not too hard to achieve (at least I hope).

**Reply by Kosette** [reply]
Thanks for your reply, #103 I saw it too, so I tried to compile it myself. But I don’t use Linux, so I’m almost clueless about what to do to compile these libs on Windows (still trying to figure it out while learning about msys2😅).

> I already succeeded in getting video to work on Windows

Yeah, this convinces me that compiling the necessary dlls on Windows is feasible, it just needs some more learning and trying from my side.

**Comment by Kosette** [comment]
If I disable some features(glimporter gluploader rawimporter), I can compile clapper.dll clapper-gtk.dll libgstclapper.dll normally, but it seems that I cannot call clappersink to play videos. 

I am not sure if it is because of incomplete compilation. 😰

**Reply by Rafostar** [reply]
Well... if it would be that easy, I wouldn't have an open issue about adding Windows support :smile: . But from what you are saying it seems most of the newly rewritten stuff compiles cleanly which is great! I just gonna find some more free time and a Windows machine to fix whatever remains.

**Comment by Rafostar** [comment]
@Kosette 

OK. I gave it a go and its caused by missing symbol exports. Its gonna take a little effort from my side, but do-able to be fixed.

**Comment by Rafostar** [comment]
@Kosette 

Work towards MS Windows support is now on `ms-win` branch with open draft PR here: https://github.com/Rafostar/clapper/pull/463

I did most of that by feeling cause I do not have Windows PC at hand at the moment. I will try compiling during week, but feel free to try it yourself and report back any remaining issues you find.

**Reply by Kosette** [reply]
Good news! Thx for your efforts.

I have compiled the latest code from the ms-win branch on GitHub action (<https://github.com/Kosette/clapper/actions/runs/9536508343/job/26283541461>), and the compilation was successful without any errors. I think this part should not have any issues.

I used the compiled dlls to recompile my project and encountered another difficulty (`configured videosink glsinkbin0 is not working` error), but I think this may be a problem dependent on my msys2 environment, and I am trying to resolve it.

For now, it seems to me that there are no issues with compiling on Windows.

**Reply by Rafostar** [reply]
Are you able to run Clapper itself on Windows and play a video? First thing to do should be getting Clapper application to work on its own.

**Reply by Rafostar** [reply]
On `mingw64` glsinkbin is packaged in [mingw-w64-gst-plugins-base](https://packages.msys2.org/base/mingw-w64-gst-plugins-base), but you should install also other related gstreamer packages, mainly `-libav` (softawer video and audio decoders), `-bad` (DirectX Hardware Acceleration), `-good` and possibly `-ugly` (stuff with license issues).

Packages list [here](https://packages.msys2.org/search?t=pkg&q=gst).

**Reply by Kosette** [reply]
> On `mingw64` glsinkbin is packaged in [mingw-w64-gst-plugins-base](https://packages.msys2.org/base/mingw-w64-gst-plugins-base), but you should install also other related gstreamer packages, mainly `-libav` (softawer video and audio decoders), `-bad` (DirectX Hardware Acceleration), `-good` and possibly `-ugly` (stuff with license issues).

Yes, I installed all gst packages. And checked `gst-inspect-1.0 glsinkbin` prints. 

> Are you able to run Clapper itself on Windows and play a video? First thing to do should be getting Clapper application to work on its own.

I tried to use clapper.exe to play directly `clapper.exe video.mp4`, the window exited immediately even before showing up. Also I tried `GST_DEBUG=3 gst-play-1.0 --videosink=clappersink 'video.mp4'`, exited with prints:
```
0:00:00.060764800  8884    227c4313970 ERROR   clapperimporterloader gstclapperimporterloader.c:162:_obtain_available_modules_once: Could not load module, reason: Error opening directory “c:\lib\clapper-0.0\gst\plugin\importers”: No such file or directory
Use Windows high-resolution clock, precision: 1 ms

Press 'k' to see a list of keyboard shortcuts.
Now playing D:\path\to\video.mp4
0:00:00.074756600  8884    227c441b190 FIXME                qtdemux qtdemux_types.c:268:qtdemux_type_get: unknown QuickTime node type .TIM
0:00:00.077513800  8884    227c441b190 FIXME                qtdemux qtdemux_types.c:268:qtdemux_type_get: unknown QuickTime node type .TSC
0:00:00.080216600  8884    227c441b190 FIXME                qtdemux qtdemux_types.c:268:qtdemux_type_get: unknown QuickTime node type .TSZ
Redistribute latency...
0:00:00.085661300  8884    227c441b000 FIXME             decodebin3 gstdecodebin3.c:1700:update_requested_selection:<decodebin3-0> Implement EXPOSE_ALL_MODE
Redistribute latency...
0:00:00.089017100  8884    227c441b2d0 FIXME             decodebin3 gstdecodebin3.c:1700:update_requested_selection:<decodebin3-0> Implement EXPOSE_ALL_MODE
.......
```
The error message at the beginning indicates that the file cannot be found, but the actual installation file path is not in c:\lib\clapper-0.0. I'm not sure how this path came about.

I installed msys2 using Scoop, and I'm not sure if this has any impact, even though I have already set the relevant environment variables.

**Reply by Rafostar** [reply]
> I have compiled the latest code from the ms-win branch on GitHub action (https://github.com/Kosette/clapper/actions/runs/9536508343/job/26283541461), and the compilation was successful

```
-Dglimporter=disabled -Dgluploader=disabled -Drawimporter=disabled
```

You should not disable them. They are needed and should compile now.

```
meson install --destdir=D:\a\clapper\clapper\clapper-dist
```
This will probably not work correctly, since meson determines importers code path at setup. If setting custom dir is necessary use `meson setup builddir --prefix=YOUR_PATH` instead.

**Reply by Kosette** [reply]
> ```
> 
> -Dglimporter=disabled -Dgluploader=disabled -Drawimporter=disabled
> 
> ```
> 
> 
> 
> You should not disable them. They are needed and should compile now.
> 

In the latest ms-win branch compilation, I have removed these configurations and kept the default options. 

These options were added a long time ago when trying to compile the master branch. 


> ```
> 
> meson install --destdir=D:\a\clapper\clapper\clapper-dist
> 
> ```
> 
> This will probably not work correctly, since meson determines importers code path at setup. If setting custom dir is necessary use `meson setup builddir --prefix=YOUR_PATH` instead.

I'll give this a try. I'm not very familiar with meson, I'm learning it while reading the documentation. 

Actually meson install is used to conveniently package the compiled artifacts. So I can package the dependencies needed to run.

**Reply by Rafostar** [reply]
> Could not load module, reason: Error opening directory “c:\lib\clapper-0.0\gst\plugin\importers”: No such file or directory

I added custom logic for this problem in latest commit. Hopefully this path will be correctly determined now.

**Reply by Kosette** [reply]
> I added custom logic for this problem in latest commit. Hopefully this path will be correctly determined now.

I'm trying to compile locally to exclude certain environmental factors. My monitor just dead, there are always unexpected circumstances🤣.

**Reply by Kosette** [reply]
> I'm trying to compile locally to exclude certain environmental factors.

@Rafostar 
I have compiled locally on my PC and encountered the following error when running with `CLAPPER_SINK_IMPORTER_PATH=/mingw64/lib/clapper-0.0/gst/plugin/importers GST_DEBUG=3 clapper.exe video.mp4`:

```
0:00:02.660626800 15464    1fcc7ababd0 FIXME              glcontext gstglcontext.c:2038:gst_gl_wrapped_context_get_config:<glwrappedcontext0> wrapped context could not retrieve config. The application may be missing a call to gst_gl_context_fill_info() or the specific platform implemention is not implemented for retrieving the config from a wrapped OpenGL context.
fish: Job 1, 'CLAPPER_SINK_IMPORTER_PATH=/min…' terminated by signal SIGSEGV (Address boundary error)
```
Did I set something wrong?

**Reply by Rafostar** [reply]
I added an env just in case, but in following commit I improved automatic path detection on Windows, so please try running without it. Just `GST_DEBUG=4 clapper.exe` and see what happens and output.

**Reply by Kosette** [reply]
Likely the same error occurs. Verbose logs https://pastebin.com/YVaykMLh

`0:00:02.989763400 13292    1e3c760c1a0 FIXME              glcontext gstglcontext.c:2038:gst_gl_wrapped_context_get_config:<glwrappedcontext0> wrapped context could not retrieve config. The application may be missing a call to gst_gl_context_fill_info() or the specific platform implemention is not implemented for retrieving the config from a wrapped OpenGL context.
`

**Reply by Rafostar** [reply]
> The error message at the beginning indicates that the file cannot be found, but the actual installation file path is not in c:\lib\clapper-0.0. I'm not sure how this path came about.

After last commit, this part now works correctly for you. Path is automatically determined and "importer" DLLs are found and loaded without problems:

```sh
0:00:02.743940900 13292    1e3c51d0a20 INFO    clapperimporterloader gstclapperimporterloader.c:137:_obtain_available_modules_once: Win32 importers path: \Path\to\msys2\mingw64\lib\clapper-0.0\gst\plugin\importers
0:00:02.749317300 13292    1e3c51d0a20 INFO    clapperimporterloader gstclapperimporterloader.c:182:_obtain_available_modules_once: Found module: libgstclapperglimporter.dll
0:00:02.753108100 13292    1e3c51d0a20 INFO    clapperimporterloader gstclapperimporterloader.c:182:_obtain_available_modules_once: Found module: libgstclappergluploader.dll
```

This message is just a "fixme" info printed. Seems to happen with surfaceless GL context, but same happens on Linux and it can be probably ignored.

```sh
0:00:02.989763400 13292    1e3c760c1a0 FIXME              glcontext gstglcontext.c:2038:gst_gl_wrapped_context_get_config:<glwrappedcontext0> wrapped context could not retrieve config. The application may be missing a call to gst_gl_context_fill_info() or the specific platform implemention is not implemented for retrieving the config from a wrapped OpenGL context.
```

The crash now happens after starting playback.

```sh
0:00:04.575872400 13292    1e3c760c880 INFO             clappersink gstclappersink.c:598:gst_clapper_sink_start:<sink> Start
fish: Job 1, 'GST_DEBUG=4 clapper 'video.mp4'' terminated by signal SIGSEGV (Address boundary error)
```

So at least everything is compiled and seems to be now correctly installed with all DLLs found and loaded. A big step forward :smile:

**Reply by Rafostar** [reply]
Could you check if you can run artifact from: https://github.com/Rafostar/clapper/actions/runs/9587958489
Just extract it to some directory and try running `clapper.exe` inside `bin` folder.

**Comment by Kosette** [comment]
@Rafostar 

I've got great news. Although it is still not possible to play normally, clapper launches successfully.

#### Clapper on Win10
![Untitled](https://github.com/Rafostar/clapper/assets/35268640/8fc4c5c5-a308-4d11-9b00-db0c5b3b5aad)

#### Error playing video (As far as I know, libgstx264.dll has been installed into lib/gstreamer-1.0 folder)
![Annotation 2024-06-19 223801](https://github.com/Rafostar/clapper/assets/35268640/f54ec2b3-2b2a-4c11-b11f-84ee1ac97bc1)

And I find another weird thing, video does not appear in file selection dialogue, to be precise, it has not been identified as a media file.

#### No video
![Annotation 2024-06-19 223858](https://github.com/Rafostar/clapper/assets/35268640/5a81abd6-a38c-4d8a-b5b9-4b0f74e67322)

#### video appeared
![Annotation 2024-06-19 223925](https://github.com/Rafostar/clapper/assets/35268640/088f3842-9886-470f-8d58-00c99d69bad3)

**Reply by Rafostar** [reply]
Does playback error continue after deleting registry file in `C:\Users\$YOUR_NAME\AppData\Local\Microsoft\Windows\INetCache\gstreamer-1.0\` ?

**Reply by Kosette** [reply]
Yes. Missing plugin error continues. And the cache bin file will always be generated.

**Reply by Rafostar** [reply]
In latest artifact build, I hope the file selection is now fixed. Also I have split binary into a normal (GUI) and console versions (`clapper-console`) to make debugging easier (i.e. `clapper-console.exe --gst-debug=2`). There is also `gst-inspect-1.0.exe` added in bin folder. An output from inspect should list all available/found plugins (including decoders). Can you upload it?

**Reply by Kosette** [reply]
File selection has been fixed.

This is `clapper-console.exe --gst-debug=2` logs: https://pastebin.com/BrNpmmwK

This is `gst-inspect-1.0.exe` logs: https://pastebin.com/iQA5jbhg

I've noticed some new error messages in logs besides plugin missing error on clapper window.

**Reply by Rafostar** [reply]
> The overwriting error message was: WGL version 1.1 is too low, need at least 3.0

Are you testing this in VM or on a real system? Do you have GPU drivers installed?

**Reply by Kosette** [reply]
> Are you testing this in VM or on a real system? Do you have GPU drivers installed?

You got the key point. I tested on a Win10 VM (My PC Monitor hasn't delivered to me yet.)

Just now, I tested it on a real Windows computer, and found errors gone🎉. The output has only a few WARN messages, and I didn't find any other exceptions.

![image](https://github.com/Rafostar/clapper/assets/35268640/96ab4f18-2480-4ff1-a135-dca2f8927080)

**Reply by Rafostar** [reply]
Great you got this to work. :+1: Can you share debug output with remaining warnings? We should try to fix as many as possible while we are still on it :smile: 

```
clapper-console.exe --gst-debug=2
```

**Reply by Kosette** [reply]
Tried play video, pause, resume, pause, seek, exit.

https://pastebin.com/p9F4DFQB

BTW, "不支持此接口" means "unsupported interface" or such.

NO outputs with `clapper-console.exe --gst-debug=2` only. It works well.

**Reply by Rafostar** [reply]
OK. Thanks.

> gst_d3d11_device_get_video_device_handle: D3D11 call failed: 0x80004002, 不支持此接口

This one would need looking into GStreamer source code why it happens, but the D3D11 seems to be working just fine afterwards anyway...

> transform could not transform video/x-raw(memory:D3D11Memory)

This is expected to happen since neither Clapper or GStreamer has currently support for importing D3D11Memory (GTK4 works with OpenGL textures). This causes worse performance (higher CPU usage), as it needs to go with a slower code path. So normal, until someone figures out and implements (either here or even better in GStreamer itself) how to render D3D11Memory into GTK4 or convert it efficiently into GL texture.

> QoS

The remaining warnings are QoS (quality of service) messages. These are normal and perfectly fine to appear occasionally. It informs about skipped frames and/or audio buffers in order to keep video and audio in sync. As long as you do not get many skipped frames during normal playback (only occasionally and when seeking) this is expected behavior.

**Reply by Rafostar** [reply]
I will also test this some more in the coming days. Hopefully MS Windows support + packages can be merged soon. Thanks for help with testing :+1:

**Reply by Kosette** [reply]
Great milestone, make Windows great again😂.

Thank you for your patience and efforts.

**Comment by Kosette** [comment]
@Rafostar 

I am encountering a strange issue when compiling a program using clapper Rust bindings on Windows mingw64.

If I set `windows_subsystem="windows"`, the compilation runs smoothly, but if I do not set this compile flag, the compilation fails with the error message `undefined reference to 'clapper_mpris_get_type'`.

I'm confused as Windows seems not support mpris feature. I don't know whether it is a clapper building issue or something else.

The error message as below, if it helps: https://github.com/tsukinaha/tsukimi/actions/runs/9760566305/job/26939707213#step:5:466

**Reply by Rafostar** [reply]
Not really a Rust expert, but one guess is that since bindings were generated on a Linux machine where MPRIS was possible they contain MPRIS stuff and its marked as available. Although normally GTypes are not called/registered unless used in code. Dunno why `windows_subsystem="windows"` would make it work through :man_shrugging: 

TIP: When building Clapper for another app, you can use `-Dclapper-app=disabled` to not build whole player and have less to package.

**Reply by Kosette** [reply]
> one guess is that since bindings were generated on a Linux machine where MPRIS was possible they contain MPRIS stuff and its marked as available. 

Thanks. I will check it out.🤔

**Reply by Kosette** [reply]
Finally, find the solution, turn on `lto=true` in Cargo.toml. It's a compiler optimization issue which is really a mind-fucking problem.🤣

**Reply by Rafostar** [reply]
Really? That's quite surprising. Glad you figured it out.
